pds_doi_core/util/doi_validator.py

```diff
@@ -0,0 +1,323 @@
+                if action != 'release':
+                    if not force_flag:
+                        logger.error(f"The title: '{doi.title}' has already been used for a DOI by lidvid:{lidvid}, status: {rows[last_row_index][col_index_to_status]} , doi: {rows[last_row_index][col_index_to_doi]} . You must use a different title.")
+                        raise DuplicatedTitleDOIException(f"The title: '{doi.title}' has already been used for a DOI by lidvid:{lidvid}, status: {rows[last_row_index][col_index_to_status]} , doi: {rows[last_row_index][col_index_to_doi]}. You must use a different title.")
+                    else:
+                        if action == 'draft':
+                            logger.warning(f"The title: '{doi.title}' has already been used for a DOI by lidvid:{lidvid}, status: {rows[last_row_index][col_index_to_status]} , doi: {rows[last_row_index][col_index_to_doi]} but action is draft and force_flag is True.")
+                else:
+                    logger.debug(f"CHECK_FIELD_TITLE_EXIST_TRUE_LIDVID_EXIST_TRUE_ACTION_IS_RELEASE: doi: {rows[last_row_index][col_index_to_doi]}, title: '{doi.title}' ")
+            else:
+                # If the same title exist but no DOI minted yet, nothing to be done.
+                logger.debug(f"The title: '{doi.title}' exist but no DOI exist with lidvid: {doi.related_identifier} in local database.  No DOI has been minted.")
+                logger.debug(f"CHECK_FIELD_TITLE_EXIST_TRUE_LIDVID_EXIST_FALSE: title:'{doi.title}'")
+        else:
+            # If no title exist, nothing to be done.
+            logger.debug(f"The title: '{doi.title}' does not yet exist in local database.")
+            logger.debug(f"CHECK_FIELD_TITLE_EXIST_FALSE: title:'{doi.title}'")
+
+        return 1
+
+    def _check_field_title_content(self, doi: Doi,action):
+        """ Check if the pds4 label is a bundle the title should contain bundle (ignoring case)
+            The same for: dataset, collection, document
+            Otherwise we raise a warning.
+        """
+
+        if doi.product_type.lower() == 'bundle' and 'bundle' not in doi.title.lower():
+            logger.warning(f"A 'bundle' product type should contain 'Bundle' in the title.")
+        elif doi.product_type.lower() == 'collection' and 'collectionb' not in doi.title.lower():
+            logger.warning(f"A 'collection' product type should contain 'Collection' in the title.")
+        elif doi.product_type.lower() == 'dataset' and 'dataset' not in doi.title.lower():
+            logger.warning(f"A 'dataset' product type should contain 'Dataset' in the title.")
+        elif doi.product_type.lower() == 'document' and 'document' not in doi.title.lower():
+            logger.warning(f"A 'document' product type should contain 'Document' in the title.")
+
+        return 1
+
+    def _check_field_lidvid_update_A(self, doi: Doi, action):
```

Comment:
I believe case A and B should be in the same method. In pseudo code that would be:

def check_update(doi):
  dois_in_dabase = get dois in database with same lidvid and doi field is not empty
   if len(dois_in_dabase)>0
    if doi.doi == None: # case A
      raise IllegalDOIActionException('...')
  elif doi.doi != dois_in_dabase[0].doi # case B, all the DOI for the same lidvid should be the same since we have this rule
     raise IllegalDOIActionException('...')

This error can not be forced. Unless I am missing something the code should not be more complicated than that except for the construction of the error message.

Reply:
I think I understand a little more about this update checking.  Hopefully the logic will be simpler when the two functions are combined into one.

Reply:
The code should be simpler now.